ZumaGame.cpp

```diff
@@ -0,0 +1,82 @@
+// code for Zuma Game
+// https://leetcode.com/problems/zuma-game/
+
+#include <iostream>
+#include <list>
+#include <vector>
+using namespace std;
+
+const int kMaxNeed = 6;
+struct Node{
+    char ch;
+    int count;
+    Node(char c, int i) : ch(c), count(i){}
+};
+void CleanBoard(list<Node> &board){
+    auto it = board.begin();
+    while (it != board.end()) {
+        if (it->count >= 3) {
+            it = board.erase(it);
+            // Ensure iterator != end(), before you use the it
+            // Ensure iterator != begin(), before invoking prev(it)
+            if (it != board.begin() &&
+                it != board.end() &&
+                it->ch == prev(it)->ch)
+            {
+                it->count += prev(it)->count;
+                board.erase(prev(it));
+            }
+        } else {
+            it++;
+        }
+    }
+}
+int DFS(list<Node> board, vector<int> &hand, int remain_in_hand) {
+    CleanBoard(board);
+    if (board.size() == 0)  return 0;
+    //cannot remove all the balls
+    if (remain_in_hand <= 0)  return kMaxNeed;
+    int min_need = kMaxNeed;
+    for (auto it = board.begin(); it != board.end(); ++it) {
+        int need = 3 - it->count;
+        if (hand[it->ch] >= need) {
+            hand[it->ch] -= need;
+            it->count += need;
+            list<Node> next_board = board;
```

Comment:
Remove, just pass in 'board'. The argument is 'pass by value', a copy is made implicitly.